source/extensions/common/wasm/wasm.cc

```diff
@@ -679,26 +664,62 @@ Word getHeaderMapSizeHandler(void* raw_context, Word type, Word result_ptr) {
   return wasmResultToWord(WasmResult::Ok);
 }
 
-// Body Buffer
-Word getRequestBodyBufferBytesHandler(void* raw_context, Word start, Word length, Word ptr_ptr,
-                                      Word size_ptr) {
+// Buffer
+Word getBufferBytesHandler(void* raw_context, Word type, Word start, Word length, Word ptr_ptr,
+                           Word size_ptr) {
+  if (type.u64_ > static_cast<uint64_t>(BufferType::MAX)) {
+    return wasmResultToWord(WasmResult::BadArgument);
+  }
   auto context = WASM_CONTEXT(raw_context);
-  auto result = context->getRequestBodyBufferBytes(start.u64_, length.u64_);
-  context->wasm()->copyToPointerSize(result, ptr_ptr.u64_, size_ptr.u64_);
+  auto buffer = context->getBuffer(static_cast<BufferType>(type.u64_));
+  if (!buffer) {
+    return wasmResultToWord(WasmResult::NotFound);
+  }
+  // NB: check for overflow.
+  if (buffer->length() < start.u64_ + length.u64_ || start.u64_ > start.u64_ + length.u64_) {
+    return wasmResultToWord(WasmResult::BadArgument);
+  }
+  uint64_t pointer = 0;
+  void* p = nullptr;
+  if (length.u64_ > 0) {
+    p = context->wasm()->allocMemory(length.u64_, &pointer);
+    if (!p) {
+      return wasmResultToWord(WasmResult::InvalidMemoryAccess);
+    }
+    buffer->copyOut(start.u64_, length.u64_, p);
+  }
+  if (!context->wasmVm()->setWord(ptr_ptr.u64_, Word(pointer))) {
+    return wasmResultToWord(WasmResult::InvalidMemoryAccess);
+  }
+  if (!context->wasmVm()->setWord(size_ptr.u64_, Word(length.u64_))) {
+    return wasmResultToWord(WasmResult::InvalidMemoryAccess);
+  }
   return wasmResultToWord(WasmResult::Ok);
 }
 
-Word getResponseBodyBufferBytesHandler(void* raw_context, Word start, Word length, Word ptr_ptr,
-                                       Word size_ptr) {
+Word getBufferStatusHandler(void* raw_context, Word type, Word length_ptr, Word flags_ptr) {
```

Comment:
I think it is a necessary API for more complex codes, e.g. transcoding where otherwise all the code would be required to track and pass the size of the buffer.  With this call it is possible to transfer control of the stream without being required to pass that additional piece of information in all cases.